From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: dt-12345 <damagetesting@proton.me>
Date: Wed, 28 Oct 2026 19:18:54 -0400
Subject: [PATCH] 10058: Add AARCH64 PAuth Extension ELF Dynamic Relocations

add support for aarch64 pauth extension dynamic relocations
 (except TLSDESC/IRELATIVE since ghidra doesn't have the non-auth versions of
 those)
---
 .../elf/relocation/AARCH64_ElfRelocationHandler.java     | 7 +++++--
 .../format/elf/relocation/AARCH64_ElfRelocationType.java | 9 ++++++++-
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java b/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java
index 3e8cb5e6e7..09672e854c 100644
--- a/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java
+++ b/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationHandler.java
@@ -78,6 +78,7 @@ public class AARCH64_ElfRelocationHandler extends
 			case R_AARCH64_P32_RELATIVE:
 				is64bit = false;
 			case R_AARCH64_RELATIVE:
+			case R_AARCH64_AUTH_RELATIVE:
 				if (elfRelocationContext.extractAddend()) {
 					addend = getValue(memory, relocationAddress, is64bit);
 				}
@@ -100,7 +101,8 @@ public class AARCH64_ElfRelocationHandler extends
 
 		switch (type) {
 			// .xword: (S+A)
-			case R_AARCH64_ABS64: {
+			case R_AARCH64_ABS64:
+			case R_AARCH64_AUTH_ABS64: {
 				newValue = (symbolValue + addend);
 				memory.setLong(relocationAddress, newValue);
 				if (symbolIndex != 0 && addend != 0 && !sym.isSection()) {
@@ -386,7 +388,8 @@ public class AARCH64_ElfRelocationHandler extends
 
 			case R_AARCH64_P32_GLOB_DAT:
 				is64bit = false;
-			case R_AARCH64_GLOB_DAT: {
+			case R_AARCH64_GLOB_DAT:
+			case R_AARCH64_AUTH_GLOB_DAT: {
 				// Corresponds to resolved local/EXTERNAL symbols within GOT
 				if (elfRelocationContext.extractAddend()) {
 					addend = getValue(memory, relocationAddress, is64bit);
diff --git a/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationType.java b/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationType.java
index 2a668a0f50..2acfe41c4f 100644
--- a/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationType.java
+++ b/Ghidra/Processors/AARCH64/src/main/java/ghidra/app/util/bin/format/elf/relocation/AARCH64_ElfRelocationType.java
@@ -230,6 +230,8 @@ public enum AARCH64_ElfRelocationType implements ElfRelocationType {
 	R_AARCH64_TLSLD_LDST128_DTPREL_LO12(572),
 	R_AARCH64_TLSLD_LDST128_DTPREL_LO12_NC(573),
 
+	R_AARCH64_AUTH_ABS64(580),
+
 	R_AARCH64_COPY(1024),			// Copy symbol at runtime. 
 	R_AARCH64_GLOB_DAT(1025),		// Create GOT entry.   
 	R_AARCH64_JUMP_SLOT(1026),		// Create PLT entry.
@@ -241,7 +243,12 @@ public enum AARCH64_ElfRelocationType implements ElfRelocationType {
 	R_AARCH64_TLS_DTPREL(1029),		// aka R_AARCH64_TLS_DTPREL64
 	R_AARCH64_TLS_TPREL(1030),		// aka R_AARCH64_TLS_TPREL64
 	R_AARCH64_TLSDESC(1031),
-	R_AARCH64_IRELATIVE(1032);
+	R_AARCH64_IRELATIVE(1032),
+
+	R_AARCH64_AUTH_RELATIVE(1041),
+	R_AARCH64_AUTH_GLOB_DAT(1042),
+	R_AARCH64_AUTH_TLSDESC(1043),
+	R_AARCH64_AUTH_IRELATIVE(1044);
 
 	public final int typeId;
 
-- 
2.45.1

